patches/server/0806-Add-root-admin-user-detection.patch

```diff
@@ -0,0 +1,68 @@
+We've seen plenty of bad/malicious plugins hit markets, and there's been a few close-calls with exploits in the past.
+Hopefully this helps mitigate some potential damage to servers, even if it is just a warning.
+
+Co-authored-by: Noah van der Aa <ndvdaa@gmail.com>
+
+diff --git a/src/main/java/io/papermc/paper/util/ServerEnvironment.java b/src/main/java/io/papermc/paper/util/ServerEnvironment.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..5ba45a071a5b7e9f2ef9f2dea7e07d74d83ccee4
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/util/ServerEnvironment.java
+@@ -0,0 +1,29 @@
++package io.papermc.paper.util;
++
++import com.sun.security.auth.module.NTSystem;
++import org.apache.commons.lang.SystemUtils;
++
++import java.io.IOException;
++import java.util.Arrays;
++
++public class ServerEnvironment {
++    private static final boolean runningAsRootOrAdmin;
++
++    static {
++        if (SystemUtils.IS_OS_WINDOWS) {
++            runningAsRootOrAdmin = Arrays.asList(new NTSystem().getGroupIDs()).contains("S-1-16-12288");
++        } else {
++            boolean isRunningAsRoot;
++            try (final var inputStream = Runtime.getRuntime().exec("id -u " + System.getProperty("user.name")).getInputStream()) {
++                isRunningAsRoot = new String(inputStream.readAllBytes()).trim().equals("0");
++            } catch (IOException ignored) {
++                isRunningAsRoot = false;
++            }
++            runningAsRootOrAdmin = isRunningAsRoot;
++        }
++    }
++
++    public static boolean userIsRootOrAdmin() {
++        return runningAsRootOrAdmin;
++    }
++}
+diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+index 1bf19965d12514dee34545235bfbadc0b74ddc8b..ed5a2edc8b07ddf9e0c4400430c9a17b5fb28216 100644
+--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
++++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+@@ -190,6 +190,16 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
+             DedicatedServer.LOGGER.warn("To start the server with more ram, launch it as \"java -Xmx1024M -Xms1024M -jar minecraft_server.jar\"");
+         }
+ 
++        // Paper start - detect running as root
++        if (io.papermc.paper.util.ServerEnvironment.userIsRootOrAdmin()) {
++            DedicatedServer.LOGGER.error("****************************");
++            DedicatedServer.LOGGER.error("YOU ARE RUNNING AS AN ADMINISTRATIVE OR ROOT USER. THIS IS NOT ADVISED.");
++            DedicatedServer.LOGGER.error("YOU ARE OPENING YOURSELF UP TO POTENTIAL RISKS WHEN DOING THIS.");
++            DedicatedServer.LOGGER.error("MALWARE, BAD PLUGINS, AND ATTACKERS WILL HAVE COMPLETE ACCESS AND CONTROL OF YOUR MACHINE.");
++            DedicatedServer.LOGGER.error("****************************");
```

Comment:
> ATTACKERS WILL HAVE COMPLETE ACCESS AND CONTROL OF YOUR MACHINE

I really don't like the wording here, I can see a lot of people thinking that this means Minecraft "hackers" will be able to access their machine, playing further into the common misunderstanding that giving out your IP address will allow players access to your computer etc.

Reply:
How does this sound?

> YOU ARE RUNNING THIS SERVER AS AN ADMINISTRATIVE OR ROOT USER. THIS IS NOT ADVISED.
> YOU ARE OPENING YOURSELF UP TO POTENTIAL RISKS WHEN DOING THIS.
> MALWARE AND MALICIOUS PLUGINS WILL HAVE COMPLETE ACCESS AND CONTROL OF YOUR MACHINE.

Reply:
Yeah, just taking out the last bit about attackers makes it much better.